### DISCUSSION TITLE
Question about Pymodbus read address

### DISCUSSION THREAD

**Discussion by Sora0224** [pymodbus-dev/pymodbus#953] [Q&A] [answered]
* Python: 3.10
* OS: Windows 10
* Pymodbus: 2.5.3
* Modbus Hardware (if used):  IDEC PLC

### Pymodbus Specific
* Server: tcp
* Client: tcp

### Description

Hi all, I'm quite new to pymodbus and I have faced few problems and tried several method but still no solution. 

So the situation here is I try to read and write data to my IDEC PLC, right now I'm able to write my coil but I can't seem to read the coil. Python error is:
1. AttributeError: 'ModbusIOException' object has no attribute 'bits'
2. Modbus Error: [Input/Output] Modbus Error: [Invalid Message] No response received, expected at least 8 bytes (0 received)

I'm curious about this, if I'm able to write to the correct coil which mean I able to read it right? but why can't I read it.

Another question, I read the PLC modbus address, some of them are like 400001, 300001, when I key in those number it show me error. is the maximum is 16 bits?
1. struct.error: 'H' format requires 0 <= number <= 65535

 I used this method with OMRON PLC and it work fine but when I changed to IDEC PLC, it doesn't work.

Thank you very much for reading.

### Code and Logs

```python
 # code and logs here.
from pymodbus.client.sync import ModbusTcpClient
import time
import datetime
import keyboard


IP_Address = "192.168.1.5"
PORT_NUM = 502
now = datetime.datetime.now()

client = ModbusTcpClient(IP_Address,PORT_NUM)
connected = client.connect()
print(connected)
if connected:
	print("PLC Connection Found",IP_Address, "at", now)
	while True:
		now = datetime.datetime.now()
		result = client.read_coils(5,10)
		client.write_coil(5,1)
		print(result.bits[5])
		time.sleep(1)
		print('done')


```

**Comment by dhoomakethu** [comment]
1. No response means  the device is not sending the response been (could be because you are missing slave address in the request) 
2. Regarding the address, pymodbus uses offset and not actual address, for e.g 40001 is offset 0 and so on. Refer to modbus specs on this.

**Reply by Sora0224** [reply]
> 1. No response means  the device is not sending the response been (could be because you are missing slave address in the request)
> 2. Regarding the address, pymodbus uses offset and not actual address, for e.g 40001 is offset 0 and so on. Refer to modbus specs on this.

Hi there, thanks for the reply. 

1. Just curious, if I'm able to write the server coils, doesn't that mean I got the right address? That's why it's kinda confuse me right now
2. So this is the modbus specs for the PLC (IDEC brand) I'm using.
![image](https://user-images.githubusercontent.com/70245039/177665731-2c4724e5-d3ec-4e9f-86c1-5157a0b15789.png)

**Reply by janiversen** [reply]
sorry for the late response.
1. Yes if you can write to the coils, then you have a valid address (might not be the one you thought it was).
2. This is the a very normal scheme (actually the one referred to in the specification), the your first holding register addresses 400001 is to addresses as 0 because we use offsets.

**Comment by janiversen** [comment, marked as answer]
Please read https://modbus.org/docs/Modbus_Application_Protocol_V1_1b3.pdf chapter 4.3 and 4.4 That should help you.
